File: C-Plus-Plus/sort/Bubble_Sort.cpp
/*Bubble Sort Algorithm is the simplest sorting algorithm in which each element is compared to its adjacent element 
and if first elememt is greater than second element ,then the swapping occurs.
The swapping occurs till the array is sorted.
*/
#include <bits/stdc++.h>			//Header file
using namespace std;	//For cin and cout
void bubblesort(int[], int);
int main()
{
	int n, arr[100], num;	//Declarartion of required variables
	cout << "Enter the number of elements of the array: ";
	cin >> n;	//Getting the number of elements of the array from the user
	cout << endl << "Enter the elements of the array to be sorted: " << "\n";
	for (int i = 0; i < n; i++)
	{
		//For loop for taking elements of the array from the user, can be space-seperated as well
		cin >> num;
		arr[i] = num;
	}
	bubblesort(arr, n);
	cout << "Array after sorting is:" << endl;
	for (int i = 0; i < n; i++)
	{
		//For loop for printing out the array after sorting
		cout << arr[i] << " ";
	}
}
void bubblesort(int arr[], int n)
{
	int temp;
	for (int i = 0; i < n - 1; i++)
	{
		bool sorted = true;	// declaring a sorted variable which will check if array is already sorted or not
		for (int j = 0; j < n - 1 - i; j++)
		{
			//Inner for loop
			if (arr[j] > arr[j + 1])
			{
				temp = arr[j + 1];	//Swapping the elements, temp is used to store the temporary variable
				arr[j + 1] = arr[j];
				arr[j] = temp;
				sorted = false;	// sorted value will be updated to false if there is swapping of elements occured 
			}
		}
		if (sorted)
		{
			break;
		}
	}
}
/*
Sample Output
Enter the number of elements of the array: 5
Enter the elements of the array to be sorted:6 5 2 3 1
Array after sorting is:
1 2 3 5 6

Time complexities:
Worst Case:O(n2)
Best Case:O(n)    
Average Case:O(n2)  

Space Complexity: Auxilary Space:O(2) because of variable temp and sorted.  
*/

File: PHP/sort/bubble_sort.php
<!-- Bubble Sort is a simplealgorithm in which each element is compared to its adjacent element 
and if first elememt is greater than second element,then the swapping occurs.
This process continues till the array is sorted. -->

<?php
function bubbleSort($arr)
{
	do
	{
	/*$isSwap checks if swapping has occured or not*/
		$isSwap = false;
		for( $i = 0; $i < count($arr) - 1; $i++ )
		{
		    /*if the next element is smaller than the current element,
		    swapping will occur*/
			if( $arr[$i] > $arr[$i + 1] )
			{
				$temp = $arr[$i];
				$arr[$i] = $arr[$i + 1];
				$arr[$i + 1] = $temp;
				/*updating the value to true as swapping occured*/
				$isSwap = true; 
			}
		}
	}
	while( $isSwap ); 
	/*returning sorted array*/
	return $arr;
}

/*declaration and initialization of $arr*/
$arr = array(4, 0, 3, 1, -1, 6);
echo 'Original array is:    ';
$i=0;
for(;$i<count($arr)-1;$i++)
	echo $arr[$i].',';
echo $arr[$i]."\n";

/*call of sort function*/
$arr = bubbleSort($arr);

/*printing the finally sorted array*/
echo 'Sorted array is:  ';
$i=0;
for(;$i<count($arr)-1;$i++)
	echo $arr[$i].',';
echo $arr[$i];

?>

<!-- Output Screen =>

Original array is:    4,0,3,1,-1,6
Sorted array is:  -1,0,1,3,4,6

Time complexity : O(n^2)
Space complexity : O(1) -->
